src/freenas/usr/local/lib/middlewared_truenas/plugins/failover.py

```diff
@@ -1373,11 +1373,41 @@ def _flush_journal(self):
             try:
                 self.middleware.call_sync('failover.call_remote', 'datastore.sql', [query, params])
             except Exception as e:
-                if isinstance(e, CallError) and e.errno in [errno.ECONNREFUSED, errno.ECONNRESET]:
-                    logger.trace('Skipping journal sync, node down')
+                if isinstance(e, CallError):
+                    if e.errno in [errno.ECONNREFUSED, errno.ECONNRESET]:
+                        logger.trace('Skipping journal sync, node down')
+                    elif e.errno == errno.EFAULT and 'sqlite3.IntegrityError' in str(e):
+                        # single-stepping the `process()` method above produced the fact
+                        # that we can hit a sqlite3.IntegrityError when trying to insert an
+                        # entry in the `storage_scrub` table.
+
+                        # If we fail here, this is where it gets gross and `/data/ha-journal`
+                        # is appended and _ALL_ datastore.* operations fail to propagate to the
+                        # standby node. The shared-memory queue object between this thread
+                        # and the `datastore` thread is cleared out. However, because the HEAD
+                        # of the `/data/ha-journal` is tried first, it fails and no other
+                        # entries get an opportunity to run so we append more entries to
+                        # `/data/ha-journal`. This is extremely painful on HA systems because,
+                        # for example, if a user disables HA (to perform maintenance) the standby
+                        # node doesn't get this request and still thinks HA is enabled between the
+                        # nodes. Ultimately, this leads to production outages/unpredictable behavior
+                        # and databases that do not match between the nodes.
+
+                        # sqlite3.IntegrityError is particulary bad so clear out the journal
+                        # and sync a fresh database to the other side.
+                        self.journal.clear()
```

Comment:
Non-empty journal was a sign that remote node is out of sync, so we continued trying to send the database if an attempt failed, and we also were able to remember that the journal is not synchronized across reboots. How is this behavior preserved now as we clear the journal?

Reply:
I can remove the `self.journal.clear()` call since by calling `failover.send_database` actually does that for us.

Reply:
We should still account for that behavior change and store some flag (e.g. `/data/ha-journal-not-synchronized`)

Reply:
If I remove the `self.journal.clear()` call, I don't see how this is a behavior change. If you remove my newly added logic and check a few lines above we actually create a oneshot alert in the event of a failure that just tells the end-user to "Manually sync to peer". Instead of alerting the user in the event of a sqlite3.IntegrityError, I send the database for them since that entry in `/data/ha-journal` will hold up all other operations.